### ISSUE TITLE
Swift 2.0 Support

### ISSUE THREAD

**Issue by rexmas** [typelift/Swiftz#240]
Doesn't seem to compile for Swift 2.0. What's the wait look like for Swift 2.0 support?

**Comment by ikesyo** [comment]
You should see `swift-develop` branch.

**Comment by pthariensflame** [comment]
@CodaFi This might be a good opportunity to switch the primary branches to the Swift 2 ones.  Do you want to do that?

**Comment by CodaFi** [comment]
I do eventually, maybe 1 more beta, but that branch is still very very experimental and in a state of flux that's mostly unsuitable for any "real" development (outside of TypeLift).  But that's me.  If the group would prefer it, I will back down.

**Comment by CodaFi** [comment]
Certainly Swiftx is stable enough that I'd feel comfortable doing this now.

**Comment by pthariensflame** [comment]
I'll leave it up to you.  I have no real opinion on this issue.

Closing.

**Comment by CodaFi** [comment]
I've changed the default in Swiftx for now.  When Swift stabilizes (and stops crashing trying to do trivial things that worked in earlier betas) I'll take this up again.

Thanks all :sparkles:

**Comment by rexmas** [comment]
Added the following to my `Podfile` but seem to get a ton of errors when I update and compile
`pod 'Swiftz', :git => 'https://github.com/typelift/Swiftz.git', :branch => 'swift-develop'`

E.g.
![screen shot 2015-08-24 at 9 40 34 pm](https://cloud.githubusercontent.com/assets/4276518/9458880/04b33182-4aa9-11e5-85d7-0c35b3f764ec.png)
and
![screen shot 2015-08-24 at 9 42 01 pm](https://cloud.githubusercontent.com/assets/4276518/9458881/086d03fc-4aa9-11e5-848b-b572085c5db2.png)

**Comment by pthariensflame** [comment]
It may have something to do with missing submodules.  Generally, @CodaFi and I use Carthage rather than CocoaPods, so I can't tell you what exactly to do.

**Comment by rexmas** [comment]
Thanks, makes sense, I'll see if I can get this to work with a custom Podfile then :+1:

**Comment by CodaFi** [comment]
It's totally the submodules.  The pod needs to track swift-develop and the new Operadics module.  The changes should be few enough you could probably make them yourself and submit a pull request.

**Comment by rexmas** [comment]
So I created an private podspec to test this out
```
Pod::Spec.new do |s|
  s.name        = "Swiftz"
  s.version     = "0.2.4"
  s.summary     = "Swiftz is a Swift library for functional programming."
  s.homepage    = "https://github.com/typelift/Swiftz"
  s.license     = { :type => "BSD" }
  s.authors     = { "CodaFi" => "devteam.codafi@gmail.com", "pthariensflame" => "alexanderaltman@me.com" }

  s.requires_arc = true
  s.osx.deployment_target = "10.9"
  s.ios.deployment_target = "8.0"
  s.source   = { :git => "https://github.com/typelift/Swiftz.git", :branch => "swift-develop", :submodules => true }
  s.source_files = "Swiftz/*.swift", "**/Swiftx/*.swift", "**/Operadics/*.swift"
end
```

But for whatever reason on every operator implementation I'm still receiving
![screen shot 2015-08-25 at 8 57 35 am](https://cloud.githubusercontent.com/assets/4276518/9471909/5d912a30-4b07-11e5-9cc8-2f62d7c9232e.png)

Anything look off to you?

**Comment by CodaFi** [comment]
That looks like it's something got lost in the Swiftx split.  How it builds on my end I'll never know.  Let me see what I can do.

**Comment by CodaFi** [comment]
OK, so I was going through the source of the thing and I don't think you're actually on `swift-develop` when you're pulling the Pod.  Try a fresh clone for me?

**Comment by CodaFi** [comment]
I'm also re-opening this until this gets resolved.

**Comment by rexmas** [comment]
`:branch => "swift-develop"` is right there in the Podfile right?

**Comment by rexmas** [comment]
And thanks for being so responsive, much appreciated :)

**Comment by rexmas** [comment]
Started from scratch and used the podspec I posted. Did a `pod update` and still getting the same errors. What's strange is Operators.swift is being included into the pod but Combinator.swift doesn't seem to recognize the existing operators. :\

![screen shot 2015-08-25 at 10 46 32 pm](https://cloud.githubusercontent.com/assets/4276518/9486383/3ea8098a-4b7b-11e5-956a-dccfaa70fae5.png)

**Comment by rexmas** [comment]
So the only way I'm able to get this to compile is if I create a `.podspec` for `Swiftx` that includes `Operadics` and then in my `.podspec` for `Swiftz` I include `s.dependency  'Swiftx'`. Then in my Podfile I include
```
pod 'Swiftx', :podspec => 'podspecs/Swiftx.podspec'
pod 'Swiftz', :podspec => 'podspecs/Swiftz.podspec'
```
Here I'm a little confused on why just directly including the Operadics source doesn't work in Swiftz like it does in Swiftx.

Anyway, from there I compile and then have to go to all the files in `Swiftz` where there's an error and alter the files to manually `import Swiftx`, which is not ideal. Makes me wonder though, assuming `Swiftx` can act as it's own module, shouldn't it always be compiled separately and therefore require `Swiftz` to have these imports anyway?

**Comment by rexmas** [comment]
Take that back, I jumped the gun because a few errors started vanishing, but that still didn't solve those operator declaration errors.

Realized however, that those operators are declared as `infix` and not `prefix` or `postfix`.
Adding the following for all operators seems to fix the remaining errors and I can likely remove the Swiftx podspec altogether once they're all added.
```swift
prefix operator • {}
postfix operator • {}
```
This seems to be missing from the codebase in `swift-develop`. Since all the operators are defined in Operadics I opened a PR there https://github.com/typelift/Operadics/pull/2

This didn't solve anything though, there's also 3 redeclared interfaces between Swiftx and Swiftz.
E.g. Swiftx - Optional.swift
![screen shot 2015-08-26 at 12 43 41 am](https://cloud.githubusercontent.com/assets/4276518/9488066/f738fa1c-4b8b-11e5-994b-3913f197dcd1.png)
Swiftz - OptionalExt.swift
![screen shot 2015-08-26 at 12 43 52 am](https://cloud.githubusercontent.com/assets/4276518/9488070/07d26020-4b8c-11e5-9b0d-8ee9c2056dd4.png)
Since `?` is just syntactic sugar for `Optional<T>` the declarations are the same. However, one implementation is using `map` while the other is using `fmap`...

If Swiftz is dependent on Swiftx then it seems like we should pick the one that's correct and add those to Swiftx and the 3 redeclared methods in `OptionalExt.swift` should simply be removed from Swiftz altogether.

If both are needed then this may indicate the need of a new operator or function name but it looks like `fmap` just returns `self.map` anyway.

**Comment by CodaFi** [comment]
See, all of this makes me think CocoaPods isn't actually checking out swift-develop.  The code you're pointing to is valid only on master.  Everything else has either been moved into different repositories or removed entirely in swift-develop.  Unless this is some source cache screwing with you, I honestly can't explain it.

**Comment by CodaFi** [comment]
So let's start over here.  Here's a diagram of the dependencies of Swiftz

```
Swiftz
  |
  \----> Swiftx - Includes by source
  |
  \----> Operadics - Included by source
  |
  \----> SwiftCheck - Linked by tests
```
Does our/your pod file describe this hierarchy?  Or does CocoaPods only allow you to express an `X links Y` relationship?

**Comment by CodaFi** [comment]
:exclamation: I know what happened here :exclamation: 

The file `Combinator.swift` was moved from Swiftz to Swiftx a little bit ago as part of the cleanup of each framework.  In the process I purposely removed those specific sections because I didn't have valid operator declarations for them pre-Operatics, but I must have forgotten to tell git to stop tracking the file.  At the same time I updated Swiftx and didn't include the new Optional extensions there because they were already covered by OptionalExt.    

So: CocoaPods downloads Swiftx and builds Optional.swift, then downloads Swiftz and attempts to build OptionalExt.swift and **crash**.  Simultaneously, it tries to look for the operator declarations for the defunct extra Combinators.swift file and can't find them because they've been removed **crash**!

I will be remedying this shortly.

**Comment by CodaFi** [comment]
The only thing standing in the way of this is a new podspec.

@rexmas?

**Comment by rexmas** [comment]
Opened a PR that includes the podspec that worked for me https://github.com/typelift/Swiftz/pull/244

**Comment by CodaFi** [comment]
Fixed by the merge of #244.

Thank you all for bearing with me :sparkles:
